File: backend/models/Appointment.ts
import mongoose, { Document, Schema } from "mongoose";

// ================= INTERFACE =================
export interface IAppointment extends Document {
  patient: mongoose.Types.ObjectId;
  doctor: mongoose.Types.ObjectId;

  patientName: string;
  phone: string;

  cccd: String,
  gender?: string;
  address?: string;
  birthDate: Date,
  specialtyId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "Departments",
  },
  doctorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "User",
  },
  medicalHistory?: string;
  allergies?: string;

  reason: string;
  date: string;
  time: string;

  imageUrl?: string;

  status: "pending" | "confirmed" | "checked_in" | "in_progress" | "cancelled" | "completed" | "missed";
  appointmentType?: "offline" | "online";
  // Fee & Payment
  originalFee?: number;       // Phí khám gốc (từ Doctor.consultationFee)
  discountAmount?: number;    // Số tiền giảm
  insuranceCoverage?: number; // % BHYT (mặc định 80 = 80%)
  insuranceNumber?: string;   // Số thẻ BHYT
  finalFee?: number;          // Phí thực tế bệnh nhân phải trả
  promotionCode?: string;     // Mã khuyến mãi đã áp dụng
  paymentMethod?: "cash" | "insurance" | "vnpay" | "momo" | "banking";
  isPaid?: boolean;

  // Pre-visit & Digital health workflow locking
  height?: number;
  weight?: number;
  bloodSugar?: number;
  preVisitQuestionnaire?: string;
  preVisitChatHistory?: any;
  isPreVisitCompleted?: boolean;
  isLocked?: boolean;
  diagnosis?: string;
  treatment?: string;
  ePrescription?: string;
}

// ================= SCHEMA =================
const appointmentSchema = new Schema<IAppointment>(
  {
    // ===== RELATION =====
    patient: {
      type: Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },

    doctor: {
      type: Schema.Types.ObjectId,
      ref: "Doctor",
      required: true,
    },

    // ===== PATIENT INFO SNAPSHOT =====
    patientName: { type: String, required: true },
    phone: { type: String, default: null },
    cccd: { type: String, default: null },
    birthDate: { type: Date, default: null },
    gender: { type: String, default: null },
    address: { type: String, default: null },
    specialtyId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Departments",
    },
    medicalHistory: { type: String, default: null },
    allergies: { type: String, default: null },

    // ===== APPOINTMENT INFO =====
    reason: { type: String, default: null },
    date: { type: String, default: null }, // YYYY-MM-DD
    time: { type: String, default: null }, // HH:mm

    // ===== IMAGE =====
    imageUrl: { type: String, default: null },

    // ===== FEE & PAYMENT =====
    originalFee: { type: Number, default: 0 },
    discountAmount: { type: Number, default: 0 },
    insuranceCoverage: { type: Number, default: 0 },  // % BHYT thanh toán
    insuranceNumber: { type: String, default: null },
    finalFee: { type: Number, default: 0 },
    promotionCode: { type: String, default: null },
    paymentMethod: {
      type: String,
      enum: ["cash", "insurance", "vnpay", "momo", "banking"],
      default: "cash",
    },
    isPaid: { type: Boolean, default: false },

    // ===== STATUS =====
    status: {
      type: String,
      enum: [
        "pending",
        "confirmed",
        "checked_in",   // Bệnh nhân đã đến phòng khám
        "in_progress",  // Đang khám
        "cancelled",
        "completed"
      ],
      default: "pending",
    },
    appointmentType: {
      type: String,
      enum: ["offline", "online"],
      default: "offline",
    },

    // ===== PRE-VISIT & DIGITAL HEALTH WORKFLOW =====
    height: { type: Number, default: null },
    weight: { type: Number, default: null },
    bloodSugar: { type: Number, default: null },
    preVisitQuestionnaire: { type: String, default: null },
    preVisitChatHistory: { type: Schema.Types.Mixed, default: null },
    isPreVisitCompleted: { type: Boolean, default: false },
    isLocked: { type: Boolean, default: false },
    diagnosis: { type: String, default: null },
    treatment: { type: String, default: null },
    ePrescription: { type: String, default: null }
  },
  {
    timestamps: true,
  }
);


/**
 * [RACE CONDITION GUARD] — Compound unique index trên (doctor, date, time)
 * để MongoDB tự động từ chối duplicate slot ngay cả khi 2 request đến đồng thời.
 * Chỉ áp dụng khi status KHÔNG phải 'cancelled'.
 * Sử dụng partial filter expression để cho phép nhiều cancelled records.
 */
appointmentSchema.index(
  { doctor: 1, date: 1, time: 1 },
  {
    unique: true,
    partialFilterExpression: {
      status: { $in: ["pending", "confirmed", "checked_in", "in_progress", "completed"] }
    },
    name: "unique_active_slot"
  }
);

// ================= MODEL =================
const Appointment = mongoose.model<IAppointment>(
  "Appointment",
  appointmentSchema
);

export default Appointment;

File: lib/models/appointment.dart
class Appointment {
  final String id;

  final String patientId;
  final String doctorId;

  // Thông tin bệnh nhân
  final String patientName;
  final String phone;
  final String gender;
  final String address;
  final String cccd; // 👈 MỚI THÊM
  final String birthDate; // 👈 MỚI THÊM
  final String patientEmail;
  final String patientAvatar;

  // Thông tin y tế
  final String medicalHistory;
  final String allergies;
  final String reason;
  final String imageUrl;

  // Thông tin lịch khám
  final String date;
  final String time;
  final String status;

  // Thông tin bác sĩ & phòng khám (Dữ liệu Populate)
  final String departmentName;
  final String doctorName;
  final String doctorAvatar;
  final String doctorSpecialty;

  // Thông tin thanh toán (Billing) 👈 MỚI THÊM
  final double originalFee;
  final double finalFee;
  final String paymentMethod;
  final bool isPaid;
  final String appointmentType;

  // Pre-visit & Digital health workflow locking
  final double height;
  final double weight;
  final double bloodSugar;
  final String preVisitQuestionnaire;
  final bool isPreVisitCompleted;
  final bool isLocked;
  final List<dynamic> preVisitChatHistory;
  final String ePrescription;
  final String diagnosis;
  final String treatment;

  Appointment({
    required this.id,
    required this.patientId,
    required this.doctorId,
    required this.patientName,
    required this.phone,
    required this.gender,
    required this.address,
    required this.cccd,
    required this.birthDate,
    required this.patientEmail,
    required this.patientAvatar,
    required this.medicalHistory,
    required this.allergies,
    required this.reason,
    required this.date,
    required this.time,
    required this.doctorName,
    required this.doctorAvatar,
    required this.imageUrl,
    required this.status,
    required this.doctorSpecialty,
    required this.departmentName,
    required this.originalFee,
    required this.finalFee,
    required this.paymentMethod,
    required this.isPaid,
    required this.appointmentType,
    this.height = 0.0,
    this.weight = 0.0,
    this.bloodSugar = 0.0,
    this.preVisitQuestionnaire = '',
    this.isPreVisitCompleted = false,
    this.isLocked = false,
    this.preVisitChatHistory = const [],
    this.ePrescription = '',
    this.diagnosis = '',
    this.treatment = '',
  });

  factory Appointment.fromJson(Map<String, dynamic> json) {
    final doctor = json['doctor'];
    final department = json['specialtyId'] ?? json['department'];

    final doctorUser = (doctor is Map) ? doctor['userId'] : null;

    // Hàm hỗ trợ parse số tiền an toàn
    double parseDouble(dynamic value) {
      if (value == null) return 0.0;
      if (value is double) return value;
      if (value is int) return value.toDouble();
      return double.tryParse(value.toString()) ?? 0.0;
    }

    return Appointment(
      id: json['_id']?.toString() ?? '',
      patientId: (json['patient'] is Map)
          ? (json['patient']['_id']?.toString() ?? '')
          : (json['patient']?.toString() ?? ''),
      doctorId: doctor is Map ? doctor['_id']?.toString() ?? '' : '',

      patientName: json['patientName'] ?? '',
      phone: json['phone'] ?? '',
      gender: json['gender'] ?? '',
      address: json['address'] ?? '',
      cccd: json['cccd'] ?? '', // 👈 Hứng dữ liệu
      birthDate: json['birthDate']?.toString() ??
          '', // 👈 Hứng dữ liệu (có thể chứa ký tự T trong ISO date)
      patientEmail: (json['patient'] is Map)
          ? (json['patient']['email']?.toString() ?? '')
          : (json['patientEmail']?.toString() ??
              (json['email']?.toString() ??
                  (json['phone'] != null && json['phone'].toString().isNotEmpty
                      ? "${json['phone']}@happyhospital.com"
                      : 'patient@happyhospital.com'))),
      patientAvatar: (json['patient'] is Map)
          ? (json['patient']['avatar']?.toString() ?? '')
          : (json['patientAvatar']?.toString() ?? ''),

      medicalHistory: json['medicalHistory'] ?? '',
      allergies: json['allergies'] ?? '',
      reason: json['reason'] ?? '',
      date: json['date'] ?? '',
      time: json['time'] ?? '',

      doctorName: (doctorUser is Map)
          ? (doctorUser['fullName'] ?? doctorUser['name'] ?? '')
          : '',
      doctorAvatar: (doctorUser is Map) ? (doctorUser['avatar'] ?? '') : '',
      doctorSpecialty: (doctor is Map) ? (doctor['specialty'] ?? '') : '',

      imageUrl: json['imageUrl'] ?? json['imageUr'] ?? '',

      departmentName: (department is Map)
          ? (department['departmentName'] ?? department['name'] ?? '')
          : '',
      status: json['status'] ?? 'pending',

      // 👈 Hứng dữ liệu thanh toán
      originalFee: parseDouble(json['originalFee']),
      finalFee: parseDouble(json['finalFee']),
      paymentMethod: json['paymentMethod'] ?? 'cash',
      isPaid: json['isPaid'] == true,
      appointmentType: json['appointmentType']?.toString() ?? 'offline',
      height: parseDouble(json['height']),
      weight: parseDouble(json['weight']),
      bloodSugar: parseDouble(json['bloodSugar']),
      preVisitQuestionnaire: json['preVisitQuestionnaire']?.toString() ?? '',
      isPreVisitCompleted: json['isPreVisitCompleted'] == true,
      isLocked: json['isLocked'] == true,
      preVisitChatHistory: json['preVisitChatHistory'] is List ? json['preVisitChatHistory'] as List : const [],
      ePrescription: json['ePrescription']?.toString() ?? '',
      diagnosis: json['diagnosis']?.toString() ?? '',
      treatment: json['treatment']?.toString() ?? '',
    );
  }

  Appointment copyWith({
    String? status,
    bool? isPaid,
    String? appointmentType,
    double? height,
    double? weight,
    double? bloodSugar,
    String? preVisitQuestionnaire,
    bool? isPreVisitCompleted,
    bool? isLocked,
    List<dynamic>? preVisitChatHistory,
    String? ePrescription,
    String? diagnosis,
    String? treatment,
  }) {
    return Appointment(
      id: id,
      patientId: patientId,
      doctorId: doctorId,
      patientName: patientName,
      phone: phone,
      gender: gender,
      address: address,
      cccd: cccd,
      birthDate: birthDate,
      patientEmail: patientEmail,
      patientAvatar: patientAvatar,
      medicalHistory: medicalHistory,
      allergies: allergies,
      reason: reason,
      date: date,
      time: time,
      doctorName: doctorName,
      departmentName: departmentName,
      doctorAvatar: doctorAvatar,
      imageUrl: imageUrl,
      status: status ?? this.status,
      doctorSpecialty: doctorSpecialty,
      originalFee: originalFee,
      finalFee: finalFee,
      paymentMethod: paymentMethod,
      isPaid: isPaid ?? this.isPaid,
      appointmentType: appointmentType ?? this.appointmentType,
      height: height ?? this.height,
      weight: weight ?? this.weight,
      bloodSugar: bloodSugar ?? this.bloodSugar,
      preVisitQuestionnaire: preVisitQuestionnaire ?? this.preVisitQuestionnaire,
      isPreVisitCompleted: isPreVisitCompleted ?? this.isPreVisitCompleted,
      isLocked: isLocked ?? this.isLocked,
      preVisitChatHistory: preVisitChatHistory ?? this.preVisitChatHistory,
      ePrescription: ePrescription ?? this.ePrescription,
      diagnosis: diagnosis ?? this.diagnosis,
      treatment: treatment ?? this.treatment,
    );
  }
}
